plugins/org.jboss.tools.openshift.common.core/src/org/jboss/tools/openshift/internal/common/core/util/CommandLocationProvider.java

```diff
@@ -0,0 +1,154 @@
+/******************************************************************************* 
+ * Copyright (c) 2015 Red Hat, Inc. 
+ * Distributed under license by Red Hat, Inc. All rights reserved. 
+ * This program is made available under the terms of the 
+ * Eclipse Public License v1.0 which accompanies this distribution, 
+ * and is available at http://www.eclipse.org/legal/epl-v10.html 
+ * 
+ * Contributors:
+ *     Red Hat Incorporated - initial API and implementation
+ ******************************************************************************/
+package org.jboss.tools.openshift.internal.common.core.util;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.concurrent.Callable;
+
+import org.apache.commons.lang.StringUtils;
+import org.eclipse.core.runtime.Platform;
+
+public class CommandLocationProvider {
+
+	// Linux command to find a command's full path
+	private static  final String LINUX_WHICH = "which ";
+	
+	// Windows command to find a command's full path
+	private static final String WINDOWS_WHERE = "where ";
+	
+	
+	private String discoveredLocation = null;
+	private boolean searchFailed = false;
+	
+	private String linuxCommand, windowsCommand;
+	private String windowsDefaultLoc, linuxDefaultLoc;
+	
+	public CommandLocationProvider(String windowsCommand, String linuxCommand,
+			String windowsDefaultLoc, String linuxDefaultLoc) {
+		this.windowsCommand = windowsCommand;
+		this.linuxCommand = linuxCommand;
+		this.linuxDefaultLoc = linuxDefaultLoc;
+		this.windowsDefaultLoc = windowsDefaultLoc;
+	}
+	
+	
+	public  String getLocation() {
+		return findLocation(-1);
+	}
+	
+	public String getLocation(int timeout) {
+		return findLocation(timeout);
+	}
+	
+	private  String findLocation(int timeout) {
+		if( discoveredLocation != null || searchFailed )
+			return discoveredLocation;
+
+		// Vagrant hasn't been discovered yet. Let's find it. 
+		if (Platform.getOS().equals(Platform.OS_WIN32)) {
+			discoveredLocation = findLocationWindows(timeout);
+		} else {
+			discoveredLocation = findLocationLinux(timeout);
+		}
+		
+		if( discoveredLocation == null ) {
+			searchFailed = true;
+		}
+		return discoveredLocation;
+	}
+	
+	private  String findLocationWindows(int timeout) {
+		if( windowsDefaultLoc != null && new File(windowsDefaultLoc).exists()) {
+			return windowsDefaultLoc;
+		}
+		String ret = searchPathWindows();
+		if( ret == null ) {
+			ret = runCommand(WINDOWS_WHERE + windowsCommand, timeout);
+		}
+		return ret;
+	}
+	private  String findLocationLinux(int timeout) {
+		if( linuxDefaultLoc != null && new File(linuxDefaultLoc).exists()) {
```

Comment:
I disagree completely. If you provide a default location AND it exists, its probably the right location. It's most likely where the binary was installed.  There's no reason to continue searching if it's already in the expected location.